### ISSUE TITLE
Predict a new query

### ISSUE THREAD

**Issue by datistiquo** [NTMC-Community/MatchZoo#684]
I already searched here. I use right now v1. Is there any sample code (I just found a broken link)? I have my trained DRRM model and want to predict ranking documents for a new query.

How is the current state in v2 to that? 

I handleld to train the modle for my own custom text data with own fast word embedding. Normally I just would predict a new query but the output are the text IDs. So for DRRM are new words ignored  which have no embedding in the dict?

Thank you very much!

**Comment by datistiquo** [comment]
I assume in v2 it is handled automatically? I need to do just preprocessor.transform(prediction_data_pack) with my new query which is in prediction_data_pack and then unpack and afterwards evaluate the model? So mapping like to histogram is done automatically?

But how do I that in v1?

**Comment by yangliuy** [comment]
@datistiquo In v1, check https://github.com/NTMC-Community/MatchZoo/blob/1.0/data/WikiQA/gen_hist4drmm.py for generating the hisogram for DRMM.

**Comment by datistiquo** [comment]
@yangliuy Thanks. I already studied this. But how does this help me to do a prediction on a new query? Because you assign embedding via document IDs and a new query has no ID. 
But is it true that real prediction is going in calc_hist of the DRMM_List_Generator? as I see there it does not need IDs at my first guess. 

I was confused by some posts here such that you need to create a new matching matrix by hand?

But am I right that you just need to handover your new query (a single one) in the config file as 
 "text1_corpus": "./data/toy_example/ranking_test/new_ query_processed.txt",
  "text2_corpus": "./data/toy_example/ranking_test/new_ query_processed.txt",

and then manually assign the word ID from the word dict to a new query?
So can I use the predict function from main.py right out of the box for a new query with no ID?


To v2: Is the way I am wrote above right?

Sorry for those naive questions.

Many Thanks! 😃

**Comment by datistiquo** [comment]
@yangliuy:  I have some trouble to figure out  how to do üprediction with new queries?! In prediction phase the get_batch  method iterates trhouh the whole relation file created from training phase. So it seems that xthe predcition here is focused on  data conatined in trainining data. Do I need to write custom code for predictions?
Could please sketch in few words the way what thinks I need tpo write "new" doing predictions?

**Comment by bwanglzu** [comment]
hi @datistiquo , when you're using matchzoo 2.0, preprocessor will do the mapping automatically, let's say:

```python
your_preprocessor = BasicPreprocessor(params. ....)
processed_train_datapack = your_preprocessor.fit_transform(your_training_data)
```

After you preprocess the training data using `your_preprocessor`, it will store the fitted parameter into the `your_preprocessor`, so you only need to use `your_preprocessor` to transform the data pack object you wanna predict.

[This](https://github.com/NTMC-Community/MatchZoo/blob/master/tutorials/data_handling.ipynb) tutorial illustrates how to create a `DataPack`.

Let's say you have your data pack named  `data_pack_to_be_predicted`, you can directly transform the data pack using

```python
processed_test_datapack = your_preprocessor.transform(data_pack_to_be_predicted)
```

The created `processed_test_datapack` is the test data you wanna make prediction of.

After this, you can use your trained model to make prediction on the `processed_test_datapack` (create batches then make prediction):

```python
pred_generator = mz.DataGenerator(processed_test_datapack, mode='point', callbacks=[hist_callback])
pred_x, pred_y = pred_generator[:]

# make prediction using your trained drmm model.
predictions = drmm_model.predict([pred_x, pred_y])
```

You can also refer to [this](https://github.com/NTMC-Community/MatchZoo/blob/master/tutorials/wikiqa/drmm.ipynb) tutorial.

**Comment by yangliuy** [comment]
@datistiquo You can refer to my answers here #426 . Again you can generate new ids for new queries or reuse old ids if they appear before. The ids of queries in testing data are already generated during the data preprocessing stage. Actually you don't need the query id to generate the predicted matching score. They are just for more convenient/efficient implementations.

**Comment by datistiquo** [comment]
@bwanglzu  @yangliuy Thanks!
But @bwanglzu for prediction you need a format required for data pack with ids and text_right toegtehr with label. I assume you artificially create this format for a single query such that you have text_left the same query and text_right you go trhough all documents? But how do you assign the labels. Can they all be set to 1 or 0?

**Comment by datistiquo** [comment]
@bwanglzu Yes. But I mean that is the poiunt for predicting new queries. How to fomrat the data pack structure? How di you do that. My above way is the only way I can think of right now...

**Comment by bwanglzu** [comment]
@datistiquo 
![image](https://user-images.githubusercontent.com/9794489/55794860-b51b4b00-5ac6-11e9-8db7-e13727f59603.png)

**Comment by datistiquo** [comment]
@bwanglzu 😄 I read this multiple times, see above:

>>But @bwanglzu for prediction you need a format required for data pack with ids and text_right toegtehr with label. I assume you artificially create this format for a single query such that you have text_left the same query and text_right you go trhough all documents? But how do you assign the labels. Can they all be set to 1 or 0?

Did you set labels artificially to 1 or 0 as labels do not play a role for prediction. I am now on the way to format it like this for ONE new query - list it in each line with ALL doc:
```

qid    text_left                                              id_right      text_right     label

1       Hey I am a new query, which doc        1               doc1              1
         is more suitable for me?
1       Hey I am a new query, which doc        
         is more suitable for me?                      2                doc2              1
1       Hey I am a new query, which doc        
         is more suitable for me?                      3                doc3               1
....                                                                 ....                 ....

1        Hey I am a new query, which doc        
         is more suitable for me?                       N               docN              1

```

Sorry for bad formatting

**Comment by bwanglzu** [comment]
@datistiquo why do you need label for prediciton? Prediction means you do not need labels, right?

**Comment by datistiquo** [comment]
I get more and more confused @bwanglzu . You said I should read the tutorial. For Prediction you need the datapack structure?? And this must have the above format?!

So mabye again, just tell how you format the datapack to do prediction with one query to all your docs?

**Comment by bwanglzu** [comment]
@datistiquo for `training` and `valuation` stage, you do need labels to train a network or evaluate how your model performance.

For testing, you do not have label. So your datapack should consist of four columns: `query id` (`id_left`), `query text` (`text_left`), `doc id` (id_right), `doc test` (`text_right`).

![image](https://user-images.githubusercontent.com/9794489/55801267-f8c98100-5ad5-11e9-8ed9-3308fb37553d.png)

And you make prediction base on this data pack.

If you have **one query** and multiple docs (let's say 1 query and 3 docs), it should be something like:

id_left    text_left    id_right    text_right
L-0        example   R-0          your-first-document-to-be-predicted
L-0        example   R-1          your-second-document-to-be-predicted
L-0        example   R-2          your-third-document-to-be-predicted

clear?

**Comment by datistiquo** [comment]
Ok it is almost like I thought. When reading tutorial you could assume labels are neccessary part of a datapack. Ok now it is clear.

Now, I will try it.


Mayn thanks for those trivial steps! 😃

**Comment by datistiquo** [comment]
@bwanglzu  I studiy the preprocessor and units since I have custom preprocessing with stopwords and stemming... I think I can write I simple function unit and call it with apply_to_text?
I just also wonder, in basic preprocessors fit method the fixedlength is not used in there but only in transform. Is there any reason why?

**Comment by datistiquo** [comment]
@bwanglzu  The prdicte output values are also negative. What is the meaning of those values? And how do I get back the accoridng document ID/Class? I am a little unexperienced with working with ranking models. I tried model.predicted_classes which does not work...
